### ISSUE TITLE
Fatal error. Internal CLR error

### ISSUE THREAD

**Issue by vitidev** [Tornhoof/SpanJson#139]
I have project .netcore3 + System.Io.Pipelines + JsonRpc(SpanJson) and I get fatal error after 14-17 action.

     An unhandled exception of type 'System.ExecutionEngineException' occurred in Unknown Module

with dump message

    The thread tried to read from or write to a virtual address for which it does not have the appropriate access

I tried to isolate issue by removing all dependies and code step by step and using System.Text.Json instead of SpanJson I found that error caused by:

     SpanJson.JsonSerializer.NonGeneric.Utf8.SerializeToArrayPool 

in JsonRpc realization. 14-16 serialization works fine, but next crash the app.

I can reproduce crash, but cannot create work short example. This code must crash, but it works

    [DataContract]
    internal class Notification
    {
        [DataMember(Name = "jsonrpc")] public string JsonRpc { get; } = "2.0";
        [DataMember(Name = "method")] public string Method { get; }
        [DataMember(Name = "params", EmitDefaultValue = false)]
        public object[] Params { get; set; }

        private Notification()
        {
        }
        public Notification(string method, object[] @params)
        {
            Method = method;
            Params = @params;
        }
    }
    class Program
    {
        static void Main(string[] args)
        {  
            for (int i = 0; i < 20; i++)
            {
                var notification = new Notification("submit", new object[] { "qQPd5Js1Q7qP9t9z8fsASK5XrdzSLMeU5H", "0F00000000000000", "0000000f", "5DD74AFE", "00066684" });
                var buffer = JsonSerializer.NonGeneric.Utf8.SerializeToArrayPool(notification);
                ArrayPool<byte>.Shared.Return(buffer.Array);
            }
            Console.WriteLine("Hello World!");
        }
    }

Then I downloaded repo and added SpanJson as project and it works. Then change to nuget again and it crash. Change again to project - it works.
The code above works always.

a-a-a-a-a, heisenbug

**Comment by Tornhoof** [comment]
If I understand you correctly you can reproduce it with your main application, but not with your small repro. Does it e.g. happen in your main app if you change your own code before it, e.g. run System.Text.Json before it?

- Try building your own nuget package of SpanJson and use that one, this will take nuget out of the equation. If that works you might actually be running in a bug of the compiler as your recompiling the package probably with a more recent version than the published nuget package was built with.

- You can try disabling tiered compilation on your main app as that is the biggest different in produced machine code between your repro and your main app (assuming you compile in Release mode with optimizations enabled).

- Are you sure that your repro code creates output with exactly the same data as in your main app?
The serialized output in your repro is 135 bytes long, if in your main app it is like 128 bytes (when it crashes) then there might be some bug in the length calculation somewhere, which then accesses the data after the end of the array (your exception). But you should receive a managed Index Exception for that error then.

- In general these kind of errors point to holes in the GC, but I couldn't explain why this should happen here.

**Comment by vitidev** [comment]
>you can reproduce it with your main application, but not with your small repro.

I can't reproduce as code snippet.  But I shrink test app and isolate to repro example.

>Try building your own nuget package of SpanJson and use that one

I compiled and replace dll and rebuild - it works, but custom nuget package does not help

>You can try disabling tiered compilation

I add `<TieredCompilation>false</TieredCompilation>` and for the first time I can see error in debugger

![Image 1](https://user-images.githubusercontent.com/2243958/69415827-01e22800-0d1e-11ea-915b-51733f04412c.png)

Can I create a private repository where to make a project with a reproducible problem?

**Comment by Tornhoof** [comment]
Yeah, sure you can do a private repo.

**Comment by vitidev** [comment]
https://github.com/vitidev/SpanJsonClrIssue/issues/1

**Comment by vitidev** [comment]
>update: custom nuget build does not help, but adding SpanJson as project works always. heisenbug

I'm a fool. Nuget build is release, but "reference as project" I have tested only debug.

"reference as project" also crash app when in release mode + checkbox "optimize code" enabled
But not crash if "optimize code" is disabled

That means the all the fault is "code optimization" or jit

**Comment by vitidev** [comment]
works [MethodImpl(MethodImplOptions.NoOptimization)] in `\SpanJson\JsonSerializer.Generics.cs`

    [MethodImpl(MethodImplOptions.NoOptimization)]
    public static ArraySegment<byte> InnerSerializeToByteArrayPool(T input)
    {
        var jsonWriter = new JsonWriter<TSymbol>(_lastSerializationSizeEstimate);
        try
        {
            Formatter.Serialize(ref jsonWriter, input);
            _lastSerializationSizeEstimate = jsonWriter.Data.Length;
            var temp = jsonWriter.Data;
            var data = Unsafe.As<TSymbol[], byte[]>(ref temp);
            return new ArraySegment<byte>(data, 0, jsonWriter.Position);
        }
        catch
        {
            jsonWriter.Dispose();
            throw;
        }
    }

but I don’t know how to go deeper

**Comment by vitidev** [comment]
but if rewritten so and prevent optimized inlining `var data = Unsafe.As<TSymbol[], byte[]>(ref temp); to `return new ArraySegment<byte>(buff, 0, jsonWriter.Position);` - it works

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static ArraySegment<byte> InnerSerializeToByteArrayPool(T input)
    {
        byte[] buff = new byte[0]; // !!!
        var jsonWriter = new JsonWriter<TSymbol>(_lastSerializationSizeEstimate);
        try
        {
            Formatter.Serialize(ref jsonWriter, input);
            _lastSerializationSizeEstimate = jsonWriter.Data.Length;
            var temp = jsonWriter.Data;
            var data = Unsafe.As<TSymbol[], byte[]>(ref temp);
            buff = data; !!!
            return new ArraySegment<byte>(buff, 0, jsonWriter.Position);
        }
        catch
        {
            jsonWriter.Dispose();
            throw new ApplicationException(buff.ToString()); !!!!
        }
    }

Is this a compiler bug?

**Comment by Tornhoof** [comment]
I'm currently looking at the dump file after running it. According to the dump the error is in TypeHandle::HasTypeEquivalence which would makes sense if that is part of the Unsafe.As.
So yes, I think this is a bug in the JIT probably.

```
AULTING_SOURCE_LINE:  f:\workspace.4\_work\1\s\src\vm\typehandle.inl

FAULTING_SOURCE_FILE:  f:\workspace.4\_work\1\s\src\vm\typehandle.inl

FAULTING_SOURCE_LINE_NUMBER:  187

FAULTING_SOURCE_CODE:  
No source found for 'f:\workspace.4\_work\1\s\src\vm\methodtable.h'


SYMBOL_NAME:  coreclr!TypeHandle::HasTypeEquivalence+7

MODULE_NAME: coreclr

IMAGE_NAME:  coreclr.dll

STACK_COMMAND:  ~14s ; .ecxr ; kb

FAILURE_BUCKET_ID:  NULL_POINTER_READ_c0000005_coreclr.dll!TypeHandle::HasTypeEquivalence

OS_VERSION:  10.0.18362.1

BUILDLAB_STR:  19h1_release

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

FAILURE_ID_HASH:  {8608e6ca-2008-7bb7-754f-0c14ce3711e5}

Followup:     MachineOwner
```
[EngineExecutionExceptionDump.zip](https://github.com/Tornhoof/SpanJson/files/3880657/EngineExecutionException.zip)

**Comment by Tornhoof** [comment]
I highly recommend opening a bug on coreclr with your private repo for reproduction and explain them exactly the same as you did here.

**Comment by vitidev** [comment]
>with your private repo for reproduction

Is error reproduce on your machine?

**Comment by Tornhoof** [comment]
Yes, on 3.0, 3.1, 5.0 Alpha. If tiered compilation ist disabled it fails on the first request. If I don't use the ArrayPool Method but Serialize directly, then everything works. I think it's a GC hole introduced by Unsafe.As and optimization/inlining.

**Comment by vitidev** [comment]
@Tornhoof 
Еhere they ask questions that I can’t answer, but you can.

**Comment by Tornhoof** [comment]
This is fixed in v3.0.1  on Nuget.
